java-checks/src/main/java/org/sonar/java/symexec/SymbolicEvaluator.java

```diff
@@ -60,112 +61,165 @@
-  }
+  private final AssignedSymbolExtractor extractor = new AssignedSymbolExtractor();
 
-  PackedStatementStates evaluateStatement(List<ExecutionState> states, StatementTree tree) {
-    return new StatementVisitor().evaluate(PackedStatementStates.instantiateWithStates(states), tree);
-  }
+  private class SymbolicEvaluatorVisitor extends BaseTreeVisitor {
+    private PackedStatementStates currentStates;
 
-  PackedStatementStates evaluateStatement(PackedStatementStates states, Tree tree) {
-    return new StatementVisitor().evaluate(states, tree);
-  }
+    private PackedStatementStates evaluateWithState(ExecutionState executionState, Tree tree) {
+      currentStates = new PackedStatementStates();
+      currentStates.addState(executionState);
+      scan(tree);
+      return currentStates;
+    }
 
-  abstract class BaseExpressionVisitor extends BaseTreeVisitor {
-    @Override
-    public final void visitBinaryExpression(BinaryExpressionTree tree) {
-      if (tree.is(Tree.Kind.CONDITIONAL_AND)) {
-        evaluateConditionalAnd(tree);
-      } else if (tree.is(Tree.Kind.CONDITIONAL_OR)) {
-        evaluateConditionalOr(tree);
-      } else if (tree.is(Tree.Kind.EQUAL_TO)) {
-        evaluateRelationalOperator(tree, SymbolicRelation.EQUAL_TO);
-      } else if (tree.is(Tree.Kind.GREATER_THAN)) {
-        evaluateRelationalOperator(tree, SymbolicRelation.GREATER_THAN);
-      } else if (tree.is(Tree.Kind.GREATER_THAN_OR_EQUAL_TO)) {
-        evaluateRelationalOperator(tree, SymbolicRelation.GREATER_EQUAL);
-      } else if (tree.is(Tree.Kind.LESS_THAN)) {
-        evaluateRelationalOperator(tree, SymbolicRelation.LESS_THAN);
-      } else if (tree.is(Tree.Kind.LESS_THAN_OR_EQUAL_TO)) {
-        evaluateRelationalOperator(tree, SymbolicRelation.LESS_EQUAL);
-      } else if (tree.is(Tree.Kind.NOT_EQUAL_TO)) {
-        evaluateRelationalOperator(tree, SymbolicRelation.NOT_EQUAL);
+    /**
+     * evaluates a condition.
+     *
+     * a condition is an expression (of type boolean), which can occur in the following places:
+     * - condition of an if statement
+     * - condition of a loop statement: do...while, for, while (but not foreach)
+     * - left operand of a logical and && and or || operators
+     * - first operand of a ternary operator
+     * conditions of switch statement are broader and not evaluated using this method.
+     *
+     * this method evaluates the given condition. if the result is either true or false,
+     * then the execution state is returned.
+     * otherwise, if the result is unknown, then the execution state is split in two branches
+     * with result TRUE and FALSE. whenever possible constraints are also set on the variables
+     * found in the condition.
+     *
+     * @param state execution state
+     * @param tree tree representing the condition
+     * @return output states
+     */
+    private List<ExecutionState> evaluateCondition(ExecutionState executionState, ExpressionTree tree) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "evaluateCondition" is 14 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)